[ToS] App crashes when ToS are missing from db

**To Reproduce**
1. Delete the `config/tos` document from a working Ground Firebase project.
2. Wipe app data on device and sign back in.

**Expected behavior**
The project list appears.

**Actual behavior**
An exception is thrown:

```
2021-11-09 18:25:05.731 8390-8390/com.google.android.gnd E/MainViewModel: Error loading terms of service from remote db
    com.google.firebase.firestore.FirebaseFirestoreException: PERMISSION_DENIED: Missing or insufficient permissions.
        at com.google.firebase.firestore.util.Util.exceptionFromStatus(Util.java:117)
        at com.google.firebase.firestore.core.EventManager.onError(EventManager.java:166)
        at com.google.firebase.firestore.core.SyncEngine.removeAndCleanupTarget(SyncEngine.java:588)
        at com.google.firebase.firestore.core.SyncEngine.handleRejectedListen(SyncEngine.java:424)
        at com.google.firebase.firestore.core.MemoryComponentProvider$RemoteStoreCallback.handleRejectedListen(MemoryComponentProvider.java:99)
        at com.google.firebase.firestore.remote.RemoteStore.processTargetError(RemoteStore.java:562)
        at com.google.firebase.firestore.remote.RemoteStore.handleWatchChange(RemoteStore.java:446)
        at com.google.firebase.firestore.remote.RemoteStore.access$100(RemoteStore.java:53)
        at com.google.firebase.firestore.remote.RemoteStore$1.onWatchChange(RemoteStore.java:176)
        at com.google.firebase.firestore.remote.WatchStream.onNext(WatchStream.java:108)
        at com.google.firebase.firestore.remote.WatchStream.onNext(WatchStream.java:38)
        at com.google.firebase.firestore.remote.AbstractStream$StreamObserver.lambda$onNext$1$AbstractStream$StreamObserver(AbstractStream.java:119)
        at com.google.firebase.firestore.remote.-$$Lambda$AbstractStream$StreamObserver$IZBnAkL3D_hQ5ofbDpSmX2lhuPo.run(Unknown Source:4)
        at com.google.firebase.firestore.remote.AbstractStream$CloseGuardedRunner.run(AbstractStream.java:67)
        at com.google.firebase.firestore.remote.AbstractStream$StreamObserver.onNext(AbstractStream.java:110)
        at com.google.firebase.firestore.remote.FirestoreChannel$1.onMessage(FirestoreChannel.java:125)
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1MessagesAvailable.runInternal(ClientCallImpl.java:658)
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1MessagesAvailable.runInContext(ClientCallImpl.java:643)
        at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
        at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:457)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
        at com.google.firebase.firestore.util.AsyncQueue$SynchronizedShutdownAwareExecutor$DelayedStartFactory.run(AsyncQueue.java:229)
        at java.lang.Thread.run(Thread.java:764)
     Caused by: io.grpc.StatusException: PERMISSION_DENIED: Missing or insufficient permissions.
        at io.grpc.Status.asException(Status.java:541)
        at com.google.firebase.firestore.util.Util.exceptionFromStatus(Util.java:115)
        at com.google.firebase.firestore.core.EventManager.onError(EventManager.java:166) 
        at com.google.firebase.firestore.core.SyncEngine.removeAndCleanupTarget(SyncEngine.java:588) 
        at com.google.firebase.firestore.core.SyncEngine.handleRejectedListen(SyncEngine.java:424) 
        at com.google.firebase.firestore.core.MemoryComponentProvider$RemoteStoreCallback.handleRejectedListen(MemoryComponentProvider.java:99) 
        at com.google.firebase.firestore.remote.RemoteStore.processTargetError(RemoteStore.java:562) 
        at com.google.firebase.firestore.remote.RemoteStore.handleWatchChange(RemoteStore.java:446) 
        at com.google.firebase.firestore.remote.RemoteStore.access$100(RemoteStore.java:53) 
        at com.google.firebase.firestore.remote.RemoteStore$1.onWatchChange(RemoteStore.java:176) 
        at com.google.firebase.firestore.remote.WatchStream.onNext(WatchStream.java:108) 
        at com.google.firebase.firestore.remote.WatchStream.onNext(WatchStream.java:38) 
        at com.google.firebase.firestore.remote.AbstractStream$StreamObserver.lambda$onNext$1$AbstractStream$StreamObserver(AbstractStream.java:119) 
        at com.google.firebase.firestore.remote.-$$Lambda$AbstractStream$StreamObserver$IZBnAkL3D_hQ5ofbDpSmX2lhuPo.run(Unknown Source:4) 
        at com.google.firebase.firestore.remote.AbstractStream$CloseGuardedRunner.run(AbstractStream.java:67) 
        at com.google.firebase.firestore.remote.AbstractStream$StreamObserver.onNext(AbstractStream.java:110) 
        at com.google.firebase.firestore.remote.FirestoreChannel$1.onMessage(FirestoreChannel.java:125) 
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1MessagesAvailable.runInternal(ClientCallImpl.java:658) 
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1MessagesAvailable.runInContext(ClientCallImpl.java:643) 
        at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37) 
        at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123) 
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:457) 
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) 
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301) 
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636) 
        at com.google.firebase.firestore.util.AsyncQueue$SynchronizedShutdownAwareExecutor$DelayedStartFactory.run(AsyncQueue.java:229) 
        at java.lang.Thread.run(Thread.java:764) 
```

Subsequent logins result in a white screen with no errors logged.

--- a/gnd/src/main/java/com/google/android/gnd/MainViewModel.java
+++ b/gnd/src/main/java/com/google/android/gnd/MainViewModel.java
@@ -41,6 +41,7 @@ import com.google.android.gnd.ui.common.SharedViewModel;
 import com.google.android.gnd.ui.home.HomeScreenFragmentDirections;
 import com.google.android.gnd.ui.signin.SignInFragmentDirections;
 import io.reactivex.Completable;
+import io.reactivex.Maybe;
 import io.reactivex.Observable;
 import java8.util.Optional;
 import javax.inject.Inject;
@@ -162,6 +163,11 @@ public class MainViewModel extends AbstractViewModel {
     return Observable.just(SignInFragmentDirections.showSignInScreen());
   }
 
+  /**
+   * @return {@link NavDirections} to {@link com.google.android.gnd.ui.home.HomeScreenFragment} if
+   *     the ToS is already accepted or if it is missing from remote config. Else returns {@link
+   *     NavDirections} to {@link com.google.android.gnd.ui.tos.TermsOfServiceFragment}.
+   */
   private Observable<NavDirections> onSignedIn() {
     hideProgressDialog();
     if (termsOfServiceRepository.isTermsOfServiceAccepted()) {
@@ -172,6 +178,7 @@ public class MainViewModel extends AbstractViewModel {
           .map(TermsOfService::getText)
           .map(text -> SignInFragmentDirections.showTermsOfService().setTermsOfServiceText(text))
           .cast(NavDirections.class)
+          .switchIfEmpty(Maybe.just(HomeScreenFragmentDirections.showHomeScreen()))
           .toObservable();
     }
   }

--- a/gnd/src/main/java/com/google/android/gnd/MainViewModel.java
+++ b/gnd/src/main/java/com/google/android/gnd/MainViewModel.java
@@ -163,11 +163,6 @@ public class MainViewModel extends AbstractViewModel {
     return Observable.just(SignInFragmentDirections.showSignInScreen());
   }
 
-  /**
-   * @return {@link NavDirections} to {@link com.google.android.gnd.ui.home.HomeScreenFragment} if
-   *     the ToS is already accepted or if it is missing from remote config. Else returns {@link
-   *     NavDirections} to {@link com.google.android.gnd.ui.tos.TermsOfServiceFragment}.
-   */
   private Observable<NavDirections> onSignedIn() {
     hideProgressDialog();
     if (termsOfServiceRepository.isTermsOfServiceAccepted()) {

--- a/gnd/src/sharedTest/java/com/google/android/gnd/FakeData.java
+++ b/gnd/src/sharedTest/java/com/google/android/gnd/FakeData.java
@@ -42,7 +42,7 @@ public class FakeData {
   public static final String LAYER_NO_FORM_COLOR = "#00ff00";
 
   public static final TermsOfService TEST_TERMS_OF_SERVICE =
-      TermsOfService.builder().setId("1").setText("Test Terms").build();
+      TermsOfService.builder().setId(TERMS_OF_SERVICE_ID).setText(TERMS_OF_SERVICE).build();
 
   public static final User TEST_USER =
       User.builder().setId("user_id").setEmail("user@gmail.com").setDisplayName("User").build();

--- a/gnd/src/sharedTest/java/com/google/android/gnd/persistence/remote/FakeRemoteDataStore.java
+++ b/gnd/src/sharedTest/java/com/google/android/gnd/persistence/remote/FakeRemoteDataStore.java
@@ -37,6 +37,7 @@ import io.reactivex.Maybe;
 import io.reactivex.Single;
 import java.util.Collections;
 import java.util.List;
+import java8.util.Optional;
 import javax.inject.Inject;
 import javax.inject.Singleton;
 
@@ -60,11 +61,7 @@ public class FakeRemoteDataStore implements RemoteDataStore {
           .setAcl(ImmutableMap.of(FakeData.TEST_USER.getEmail(), "contributor"))
           .build();
 
-  private final TermsOfService testTermsOfService =
-      TermsOfService.builder()
-          .setId(FakeData.TERMS_OF_SERVICE_ID)
-          .setText(FakeData.TERMS_OF_SERVICE)
-          .build();
+  private Optional<TermsOfService> termsOfService = Optional.of(FakeData.TEST_TERMS_OF_SERVICE);
 
   private final Project testProjectWithNoLayers =
       Project.newBuilder()
@@ -111,9 +108,13 @@ public class FakeRemoteDataStore implements RemoteDataStore {
     return Single.just(getTestProject());
   }
 
+  public void setTermsOfService(Optional<TermsOfService> termsOfService) {
+    this.termsOfService = termsOfService;
+  }
+
   @Override
   public @Cold Maybe<TermsOfService> loadTermsOfService() {
-    return Maybe.just(testTermsOfService);
+    return termsOfService.isEmpty() ? Maybe.empty() : Maybe.just(termsOfService.get());
   }
 
   @Override

--- a/gnd/src/test/java/com/google/android/gnd/MainViewModelTest.java
+++ b/gnd/src/test/java/com/google/android/gnd/MainViewModelTest.java
@@ -22,6 +22,7 @@ import static com.google.common.truth.Truth.assertThat;
 
 import android.content.SharedPreferences;
 import androidx.navigation.NavDirections;
+import com.google.android.gnd.persistence.remote.FakeRemoteDataStore;
 import com.google.android.gnd.repository.TermsOfServiceRepository;
 import com.google.android.gnd.repository.UserRepository;
 import com.google.android.gnd.system.auth.FakeAuthenticationManager;
@@ -33,6 +34,7 @@ import com.google.android.gnd.ui.signin.SignInFragmentDirections;
 import dagger.hilt.android.testing.HiltAndroidTest;
 import io.reactivex.observers.TestObserver;
 import java.util.NoSuchElementException;
+import java8.util.Optional;
 import javax.inject.Inject;
 import org.junit.Before;
 import org.junit.Test;
@@ -45,6 +47,7 @@ import org.robolectric.shadows.ShadowToast;
 public class MainViewModelTest extends BaseHiltTest {
 
   @Inject FakeAuthenticationManager fakeAuthenticationManager;
+  @Inject FakeRemoteDataStore fakeRemoteDataStore;
   @Inject MainViewModel viewModel;
   @Inject Navigator navigator;
   @Inject SharedPreferences sharedPreferences;
@@ -136,6 +139,20 @@ public class MainViewModelTest extends BaseHiltTest {
   }
 
   @Test
+  public void testSignInStateChanged_onSignedIn_whenTosMissing() {
+    tosRepository.setTermsOfServiceAccepted(false);
+    fakeRemoteDataStore.setTermsOfService(Optional.empty());
+
+    fakeAuthenticationManager.setUser(TEST_USER);
+    fakeAuthenticationManager.signIn();
+
+    verifyProgressDialogVisible(false);
+    verifyNavigationRequested(HomeScreenFragmentDirections.showHomeScreen());
+    verifyUserSaved();
+    assertThat(tosRepository.isTermsOfServiceAccepted()).isFalse();
+  }
+
+  @Test
   public void testSignInStateChanged_onSignInError() {
     setupUserPreferences();
 
@@ -147,20 +164,5 @@ public class MainViewModelTest extends BaseHiltTest {
     verifyUserPreferencesCleared();
     verifyUserNotSaved();
     assertThat(tosRepository.isTermsOfServiceAccepted()).isFalse();
-  }
-
-  @Test
-  public void testSignInStateChanged_onSignedIn_whenTosMissing() {
-    when(mockTosRepository.isTermsOfServiceAccepted()).thenReturn(false);
-    when(mockUserRepository.saveUser(any(User.class))).thenReturn(Completable.complete());
-    when(mockTosRepository.getTermsOfService()).thenReturn(Maybe.empty());
-
-    fakeAuthenticationManager.setUser(TEST_USER);
-    fakeAuthenticationManager.signIn();
-
-    assertProgressDialogVisible(false);
-    assertNavigate(HomeScreenFragmentDirections.showHomeScreen());
-    Mockito.verify(mockUserRepository, times(1)).saveUser(TEST_USER);
-    Mockito.verify(mockTosRepository, times(0)).setTermsOfServiceAccepted(anyBoolean());
   }
 }

--- a/gnd/src/test/java/com/google/android/gnd/MainViewModelTest.java
+++ b/gnd/src/test/java/com/google/android/gnd/MainViewModelTest.java
@@ -146,4 +146,19 @@ public class MainViewModelTest extends BaseHiltTest {
     Mockito.verify(mockUserRepository, times(1)).clearUserPreferences();
     Mockito.verify(mockTosRepository, times(1)).setTermsOfServiceAccepted(false);
   }
+
+  @Test
+  public void testSignInStateChanged_onSignedIn_whenTosMissing() {
+    when(mockTosRepository.isTermsOfServiceAccepted()).thenReturn(false);
+    when(mockUserRepository.saveUser(any(User.class))).thenReturn(Completable.complete());
+    when(mockTosRepository.getTermsOfService()).thenReturn(Maybe.empty());
+
+    fakeAuthenticationManager.setUser(TEST_USER);
+    fakeAuthenticationManager.signIn();
+
+    assertProgressDialogVisible(false);
+    assertNavigate(HomeScreenFragmentDirections.showHomeScreen());
+    Mockito.verify(mockUserRepository, times(1)).saveUser(TEST_USER);
+    Mockito.verify(mockTosRepository, times(0)).setTermsOfServiceAccepted(anyBoolean());
+  }
 }

--- a/gnd/src/test/java/com/google/android/gnd/repository/TermsOfServiceRepositoryTest.java
+++ b/gnd/src/test/java/com/google/android/gnd/repository/TermsOfServiceRepositoryTest.java
@@ -21,7 +21,9 @@ import static com.google.common.truth.Truth.assertThat;
 import com.google.android.gnd.BaseHiltTest;
 import com.google.android.gnd.FakeData;
 import com.google.android.gnd.model.TermsOfService;
+import com.google.android.gnd.persistence.remote.FakeRemoteDataStore;
 import dagger.hilt.android.testing.HiltAndroidTest;
+import java8.util.Optional;
 import javax.inject.Inject;
 import org.junit.Test;
 import org.junit.runner.RunWith;
@@ -31,10 +33,13 @@ import org.robolectric.RobolectricTestRunner;
 @RunWith(RobolectricTestRunner.class)
 public class TermsOfServiceRepositoryTest extends BaseHiltTest {
 
+  @Inject FakeRemoteDataStore fakeRemoteDataStore;
   @Inject TermsOfServiceRepository termsOfServiceRepository;
 
   @Test
   public void testGetTermsOfService() {
+    fakeRemoteDataStore.setTermsOfService(Optional.of(FakeData.TEST_TERMS_OF_SERVICE));
+
     termsOfServiceRepository
         .getTermsOfService()
         .test()
@@ -43,6 +48,13 @@ public class TermsOfServiceRepositoryTest extends BaseHiltTest {
                 .setId(FakeData.TERMS_OF_SERVICE_ID)
                 .setText(FakeData.TERMS_OF_SERVICE)
                 .build());
+  }
+
+  @Test
+  public void testGetTermsOfService_whenMissing_doesNotThrowException() {
+    fakeRemoteDataStore.setTermsOfService(Optional.empty());
+
+    termsOfServiceRepository.getTermsOfService().test().assertNoValues().assertComplete();
   }
 
   @Test

